js/userDatabase.js

```diff
@@ -0,0 +1,118 @@
+          gUserDatabase._logError(event, "Cannot open an indexedDB connection");
+          reject(new Error(event.target.errorCode));
+        };
+        request.onsuccess = event => {
+          gUserDatabase._onDatabaseOpenSuccess(event).then(resolve);
+        };
+        request.onupgradeneeded = event => {
+          gUserDatabase._onDatabaseUpgrade(event);
+        };
+      });
+    },
+
+    save(objectStoreToWrite, objectStoreType, data, mockObjectStore) {
+      return new Promise((resolve, reject) => {
+        var transaction = gUserDatabase._database.transaction([objectStoreToWrite], "readwrite");
+        var objectStore = mockObjectStore || transaction.objectStore(objectStoreToWrite);
+
+        var request = objectStore.get(objectStoreType);
+        request.onsuccess = () => {
+          gUserDatabase._onWriteFetchRequestSuccess(request, data, objectStore, objectStoreType).then(resolve, reject);
+        };
+        request.onerror = event => {
+          gUserDatabase._logError(event, "Failed to store object of type " + objectStoreType);
+          reject(new Error(event.target.errorCode));
+        };
+      });
+    },
+
+    load(objectStoreToRead, objectStoreType) {
+      return new Promise((resolve, reject) => {
+        var transaction = gUserDatabase._database.transaction([objectStoreToRead]);
+        var objectStore = transaction.objectStore(objectStoreToRead);
+        var request = objectStore.get(objectStoreType);
+        var transactionDescription = "Load data " + objectStoreType;
+        gUserDatabase._setSimpleRequestHandlers(request, transactionDescription).then(resolve, reject);
+      });
+    },
+
+    _logError(event, errorString) {
+      var error = "Error: " + event.target.errorCode + ": " + errorString;
+      console.error(error);
+    },
+
+    _setSimpleRequestHandlers(request, logString) {
+      return new Promise((resolve, reject) => {
+        request.onerror = event => {
+          gUserDatabase._logError(event, logString);
+          reject(new Error(event.target.errorCode));
+        };
+        request.onsuccess = event => {
+          resolve(event.target.result.data);
+        };
+      });
+    },
+
+    _createPrefsData(dataType, data) {
+      var prefsData = {};
+      prefsData.prefType = dataType;
+      prefsData.data = data;
+      return prefsData;
+    },
+
+    _onWriteFetchRequestSuccess(request, dataToWrite, objectStore, objectStoreType) {
+      var result = request.result;
+      result.data = dataToWrite;
+      var requestUpdate = objectStore.put(result);
+      var transactionDescription = "Update data " + objectStoreType;
+      return gUserDatabase._setSimpleRequestHandlers(requestUpdate, transactionDescription);
+    },
+
+    _onDatabaseOpenSuccess(event) {
+      // Save the database connection and pass back the existing pinned links.
+      gUserDatabase._database = event.target.result;
+      return gUserDatabase.load(OBJECT_STORE_PREFS, PINNED_LINKS_PREF);
```

Comment:
I don't think we should do this here. The gUserDatabase should not know anything about pinned links. 

If PinnedLinks want to load stuff, it can ask the database to load those things - but gUserDatabase should not do this on its own.